keel-web/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -115,45 +116,60 @@ class ApplicationService(
     val environmentSummaries = getEnvironmentSummariesFor(application)
 
     return deliveryConfig.artifacts.map { artifact ->
-      repository.artifactVersions(artifact).map { version ->
+      val artifactVersionSummaries = repository.artifactVersions(artifact).map { version ->
         val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
 
         environmentSummaries.forEach { environmentSummary ->
           val environment = deliveryConfig.environments.find { it.name == environmentSummary.name }!!
           environmentSummary.getArtifactPromotionStatus(artifact, version)?.let { status ->
-            if (status == PromotionStatus.PENDING) {
-              ArtifactSummaryInEnvironment(
+            var artEnvSummary = when (status) {
+              PENDING -> ArtifactSummaryInEnvironment(
                 environment = environmentSummary.name,
                 version = version,
                 state = status.name.toLowerCase()
               )
-            } else {
-              repository.getArtifactSummaryInEnvironment(
+              SKIPPED -> {
+                // some environments contain relevant info for skipped artifacts, so
+                // try and find that before defaulting to less information
+                val potentialSummary = repository.getArtifactSummaryInEnvironment(
+                  deliveryConfig = deliveryConfig,
+                  environmentName = environmentSummary.name,
+                  artifactName = artifact.name,
+                  artifactType = artifact.type,
+                  version = version
+                )
+                if (potentialSummary == null || potentialSummary.state == "pending") {
+                  ArtifactSummaryInEnvironment(
+                    environment = environmentSummary.name,
+                    version = version,
+                    state = status.name.toLowerCase()
+                  )
+                } else {
+                  potentialSummary
+                }
+              }
```

Comment:
Struggling a bit to understand how this is different from the default case handled by the `else` block just below. Can you please explain?

Reply:
Sure - is the comment on it not clear? Suggestions on how I could rephrase?

"Skipped" artifacts are a little weird - sometimes when you get a summary for them from the database it returns "pending". That's because when we have no information about a version we return "pending" status (for example, the version was skipped in envA, which envB depends-on, we would have no info about if it was skipped in envB so we'd return a "pending" summary). 

But, if we already know it's skipped, that's not correct . That means that it the info doesn't correctly get added to the summary. So, first we check if the summary exists correctly in the database (this would be the case where there's a "skipped" entry for the env, like if there were two artifacts up for manual judgement but we only approved the latest version in that environment). If that doesn't exist, we give a summary with minimal information (like we do for pending).

Reply:
An example of this situation is in the tests for this - `ApplicationServiceTests` `ln360` shows that in the first env we have data about which version replaced the version that was skipped, but the following two environments don't have that information.